### PR TITLE
[SPARK-44141] Removed need to have buf preinstalled

### PR DESCRIPTION
**Pull request by arnarpall**
JIRA:  https://issues.apache.org/jira/browse/SPARK-44141

### What changes were proposed in this pull request?
In order to lower the barrier of entry even further for this project we can remove need to have `buf` preinstalled and just use `go run`

This also ensures that the tool chain remains consistent and there is less works on my machine issues to be had.

### Why are the changes needed?
Not really needed but will make the barrier of entry even lower

### How was this patch tested?
Run `make gen` on a completely fresh system

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -37,6 +37,9 @@ GOARCH                    ?= $(shell go version | cut -d' ' -f4 | cut -d'/' -f2)
 TAGS                      ?= netgo
 SHELL = bash
 
+## Build tools
+BUF                       := $(GO) run github.com/bufbuild/buf/cmd/buf@v1.21.0
```
Comment (grundprinzip) [inline Makefile:41]: Thats very nice! Do you mind changing as well the github workflow?

**Comment by arnarpall** [comment]
> Thanks for your contributions, can you please open a Spark Jira for this change?

Yeah absolutely. I've requested an account and will create an issue as soon as the account is ready.

**Review by grundprinzip — APPROVED** [review]
@HyukjinKwon can you please approve the workflow for @arnarpall? Thanks!

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -47,15 +47,8 @@ jobs:
         run: |
           curl -LO https://go.dev/dl/go1.19.9.linux-amd64.tar.gz
           sudo tar -C /usr/local -xzf go1.19.9.linux-amd64.tar.gz
-      - name: Install Buf
```
Comment (HyukjinKwon) [inline .github/workflows/build.yml:50]: Hm, seems the build fails 😢

Comment (arnarpall) [reply .github/workflows/build.yml:50]: I'll see if I can take a look tomorrow.

**Comment by arnarpall** [comment]
@HyukjinKwon  hey hey. I finally managed to get back to this PR, could you perhaps approve the workflow for me ?

**Comment by arnarpall** [comment]
I seems like not all the changes to the workflow are not being reflected properly.

The current failure.
Output from the `internal/generated.out` target is for the run
```
>> BUILD, output = internal/generated.outbuf generate --debug -vvv
bash: line 1: buf: command not found
```

However this is what we should be expecting 
```
>> BUILD, output = internal/generated.outGO111MODULE=on go run github.com/bufbuild/buf/cmd/buf@v1.26.1 generate --debug -vvv
```
Here instead of calling the buf binary directly we expect it to be wrapped in a `go run` statement

**Comment by grundprinzip** [comment]
Thanks a lot for your contribution. I have fixed the merge conflict and verified that it works correctly now!

Will merge to master once the PR is green.

**Pull request closed without merging**